src/dataset.py

```diff
@@ -1,24 +1,94 @@
+    def __init__(self, dataset: np.array, labels_index: int):
         self._data = dataset
+        self._labels_index = labels_index
 
-    def load_from_csv(self, path) -> 'Dataset':
-        pass
+    # Idea for the loading data: creating another module which allows to generalize the format of the data imported
+    # Using for example an abstract class for the loading package.
 
-    def load_from_array(self, data: np.array) -> 'Dataset':
-        pass
+    @staticmethod
+    def load_from_csv(path: str, csv_limiter: str, label_name: str, new_line: str = '') -> 'Dataset':
 
+        with open(path, newline=new_line) as csvfile:
+            data = []
+            reader = csv.reader(csvfile, delimiter=csv_limiter)
+            for row in reader:
+                data.append(row)
 
-    def chose_best_split(self):
-        # TODO : This function contains the algo which allows us to take the split from the dataset. I have to think about the link of a condition
-        pass
+            label_index = data[0].index(label_name)
+            data = np.asarray(data)
+            data = data[1:]
 
-    def split_condition(self, condition):
-        # TODO : This function allows us to split the dataset given a condition. The format of the conditon is still unknown
-        pass
+        dataset = Dataset(dataset=data, labels_index=label_index)
+        return dataset
+
+    @staticmethod
+    def load_from_array(data: np.array, label_index: int) -> 'Dataset':
+        dataset = Dataset(dataset=data, labels_index=label_index)
+        return dataset
+
+    def _get_possible_split(self):
+        all_possibilites = {}
+
+        for i in range(self._data.shape[0]):
+            if i == self._labels_index:
+                continue
+
+            all_possibilites[i] = np.unique(self._data[:, i])
+        return all_possibilites
+
+    def _best_split_for_each_feature(self, purity_evaluation: Information_gain):
+        ig_gain_by_feature = {}
+        possibilites = self._get_possible_split()
 
+        for i in range(self._data.shape[0]):
+            ig_gain_max = -np.inf
+            best_value = None
+            best_feature_index = None
+            if i == self._labels_index:
+                continue
+
+            for possibility in possibilites[i]:
+                mask = self._get_condition()
+                ig_gain_new = purity_evaluation.purity_value(labels_values=self._data[:, self._labels_index], mask=mask)
+                if ig_gain_new > ig_gain_max:
+                    ig_gain_max = ig_gain_new
+                    best_value = possibility
+                    best_feature_index = i
+
+            ig_gain_by_feature[i] = [ig_gain_max, best_value, best_feature_index]
+        return ig_gain_by_feature
+
+    def chose_best_split(self, purity_evaluation: Information_gain):
+        ig_dict = self._best_split_for_each_feature(purity_evaluation)
+        max_ig = 0
+        split = None
+        for value_ig in ig_dict.values():
```

Comment:
I find the name value_ig confusing. I thought it was the information gain value, but it seems you are storing the split value in it? i.e. split has an information gain and the split occurs at a value of the particular feature.
Would rename the overall variable to be more clear